File: common/src/main/java/io/xj/lib/music/Root.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.
package io.xj.lib.music;

import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 Root can be the root of a Chord, Key or Scale.
 */
public class Root {
  private static final Pattern rgxNote = Pattern.compile("^([ABCDEFG]).*");
  private static final Pattern rgxNoteModified = Pattern.compile("^([ABCDEFG][♯#♭b]).*");
  private static final Pattern rgxSlashNote = Pattern.compile("/([ABCDEFG])$");
  private static final Pattern rgxSlashNoteModified = Pattern.compile("/([ABCDEFG][♯#♭b])$");
  private PitchClass pitchClass;
  private String remainingText;

  /**
   Parse root and remaining string, using regular expressions

   @param name to parse root and remaining of
   */
  private Root(String name) {
    // as a default, the whole thing is remaining text, and pitch class is None
    this.pitchClass = PitchClass.None;
    this.remainingText = name;

    evaluate(rgxNote, name, true);
    evaluate(rgxNoteModified, name, true);
    evaluate(rgxSlashNote, name, false);
    evaluate(rgxSlashNoteModified, name, false);
  }

  /**
   Instantiate a Root by name
   <p>
   [#176728338] XJ understands the root of a slash chord

   @param name of root
   @return root
   */
  public static Root of(String name) {
    return new Root(name);
  }

  /**
   First group matching pattern in text, else null@param pattern to in

   @param text              to search
   @param withRemainingText whether to store remaining text from this evaluation
   */
  private void evaluate(Pattern pattern, String text, Boolean withRemainingText) {
    Matcher matcher = pattern.matcher(text);
    if (!matcher.find())
      return;

    String match = matcher.group(1);
    if (Objects.isNull(match) || match.length() == 0)
      return;

    this.pitchClass = PitchClass.of(match);
    if (withRemainingText)
      this.remainingText = text.substring(match.length()).trim();
  }

  /**
   Get pitch class of root

   @return root pitch class
   */
  public PitchClass getPitchClass() {
    return pitchClass;
  }

  /**
   Remaining text after root has been extracted

   @return remaining text
   */
  public String getRemainingText() {
    return remainingText;
  }

}

File: engine/src/music/Root.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include "xjmusic/music/Root.h"

using namespace XJ;


std::regex Root::rgxNote("^([ABCDEFG]).*");
std::regex Root::rgxNoteModified("^([ABCDEFG][#b]).*");


Root::Root(const std::string &name) {
  {
    const std::string normalized = accidentalNormalized(name);

    // as a default, the whole thing is remaining text, and pitch class is None
    pitchClass = PitchClass::Atonal;
    remainingText = normalized;

    evaluate(rgxNote, normalized);
    evaluate(rgxNoteModified, normalized);
  }
}


Root Root::of(const std::string &name) {
  return Root(name);
}


void Root::evaluate(const std::regex &pattern, const std::string &text) {
  std::smatch matcher;
  if (!std::regex_search(text, matcher, pattern))
    return;

  const std::string match = matcher[1];
  if (match.empty())
    return;

  this->pitchClass = pitchClassOf(match);
  this->remainingText = text.substr(match.length());
  this->remainingText.erase(this->remainingText.begin(),
                            std::find_if(this->remainingText.begin(), this->remainingText.end(), [](const unsigned char ch) {
                              return !std::isspace(ch);
                            }));
}
